indexer/osm_editor.cpp

```diff
@@ -678,21 +690,31 @@ NewFeatureCategories Editor::GetNewFeatureCategories() const
   return res;
 }
 
-namespace
+FeatureID Editor::GenerateNewFeatureId(MwmSet::MwmId const & id)
 {
-FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
-{
-  // TODO(AlexZ): Stable & unique features ID generation.
-  // TODO(vng): Looks like new feature indexes should uninterruptedly continue after existing indexes in mwm file.
-  static uint32_t newIndex = 0x0effffff;
-  return FeatureID(oldFeatureId.m_mwmId, newIndex++);
+  // TODO(vng): Double-check if new feature indexes should uninterruptedly continue after existing indexes in mwm file.
+  uint32_t featureIndex = kStartIndexForCreatedFeatures;
+  auto const found = m_features.find(id);
+  if (found != m_features.end())
+  {
+    // Scan all already created features and choose next available ID.
+    for (auto const & feature : found->second)
+    {
+      if (feature.second.m_status == FeatureStatus::Created && featureIndex <= feature.first)
+        featureIndex = feature.first + 1;
```

Comment:
Don’t see any issue with free slots initialization. We have a plenty of free IDs available (and their number can be increased even more).
And no atomics are needed.

> On Mar 9, 2016, at 12:34, ygorshenin <notifications@github.com> wrote:
> 
> In indexer/osm_editor.cpp <https://github.com/mapsme/omim/pull/2106#discussion_r55492062>:
> 
> > -FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
> > -{
> > -  // TODO(AlexZ): Stable & unique features ID generation.
> > -  // TODO(vng): Looks like new feature indexes should uninterruptedly continue after existing indexes in mwm file.
> > -  static uint32_t newIndex = 0x0effffff;
> > -  return FeatureID(oldFeatureId.m_mwmId, newIndex++);
> > +  // TODO(vng): Double-check if new feature indexes should uninterruptedly continue after existing indexes in mwm file.
> > +  uint32_t featureIndex = kStartIndexForCreatedFeatures;
> > +  auto const found = m_features.find(id);
> > +  if (found != m_features.end())
> > +  {
> > +    // Scan all already created features and choose next available ID.
> > +    for (auto const & feature : found->second)
> > +    {
> > +      if (feature.second.m_status == FeatureStatus::Created && featureIndex <= feature.first)
> > +        featureIndex = feature.first + 1;
> Okay, I'll propose you a better approach:
> 1. Create an atomic variable.
> 2. add_and_fetch() on each GenerateNewFeatureId().
> 
> This is simpler and thread-safe, right? (joke)
> You do not utilize free slots here, whereas you can simply do it.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/mapsme/omim/pull/2106/files#r55492062>.
>